executor/adapter.go

```diff
@@ -481,22 +481,48 @@ func (a *ExecStmt) handleNoDelay(ctx context.Context, e Executor, isPessimistic
 	}
 
 	// If the executor doesn't return any result to the client, we execute it without delay.
-	if toCheck.Schema().Len() == 0 {
+	if isNoResultExecutor(toCheck) {
 		handled = !isExplainAnalyze
 		if isPessimistic {
 			return handled, nil, a.handlePessimisticDML(ctx, toCheck)
 		}
 		r, err := a.handleNoDelayExecutor(ctx, toCheck)
 		return handled, r, err
-	} else if proj, ok := toCheck.(*ProjectionExec); ok && proj.calculateNoDelay {
+	}
+
+	return false, nil, nil
+}
+
+func isNoResultExecutor(e Executor) bool {
+	if e.Schema().Len() == 0 {
+		return true
+	}
+	if proj, ok := e.(*ProjectionExec); ok && proj.calculateNoDelay {
```

Comment:
This is different from before. It didn't call `handlePessimisticDML` in this case. Are you intended to modify the logic?

Reply:
I'm not intending to modify the old logic... just a tiny refactor ...
I get trapped here because this special logic, and want to avoid others been trapped again, so the refactor

Reply:
OK, I just reset the change here and touch only the necessary code.